contract/src/main/kotlin/io/provenance/scope/loan/utility/ContractRequirements.kt

```diff
@@ -23,55 +28,68 @@ internal infix fun Boolean.orError(error: ContractViolation): ContractEnforcemen
+            }
+        }
+    }.handleViolations()
 
 /**
  * Performs validation of any [ContractEnforcement]s specified by [requireThat] in the body.
  * Should be used to wrap code blocks that also performs actions other than validation.
  */
 internal fun validateRequirements(
-    checksBody: ContractViolationMap.() -> Unit
-) = mutableMapOf<ContractViolation, Int>()
+    checksBody: ContractEnforcementContext.() -> Unit
+) = ContractEnforcementContext()
     .apply(checksBody)
     .handleViolations()
 
 /**
- * Performs validation of one or more [ContractEnforcement]s.
- */
-internal fun validateRequirements(
-    vararg checks: ContractEnforcement
-) = checks.fold<ContractEnforcement, ContractViolationMap>(emptyMap()) { acc, (rule, violationReport) ->
-        if (!rule) {
-            acc.plus(violationReport to (acc.getOrDefault(violationReport, 0) + 1))
-        } else {
-            acc
-        }
-    }.handleViolations()
-
-/**
- * Aggregates multiple [ContractEnforcement]s into a single exception listing all of the [ContractViolation]s that were
- * found.
+ * Aggregates multiple [ContractEnforcement]s into a single exception listing all of the [ContractViolation]s that
+ * were found.
  *
- * @throws [IllegalArgumentException] if any of the supplied contract requirements were violated.
+ * @throws [ContractViolationException] if any of the supplied contract requirements were violated.
  * @return [Nothing] if no contract requirements were violated.
  */
 private fun ContractViolationMap.handleViolations() =
-    takeIf { map ->
-        map.isNotEmpty()
-    }?.let { violations -> // TODO: Determine how to best list out all the exceptions as a string
-        violations.entries.joinToString { (violation, count) ->
-            if (count > 1) {
-                "$violation ($count occurrences)"
-            } else {
-                violation
-            }
+    entries.fold(
+        "" to 0U
+    ) { (violationMessage, overallViolationCount), (violation, count) ->
+        if (count > 0U) {
+            "$violationMessage; $violation ($count occurrences)" to overallViolationCount + 1U
+        } else {
+            violationMessage to overallViolationCount
         }
-    }?.let { formattedViolations ->
-        throw IllegalArgumentException("The contract input was invalid: $formattedViolations")
+    }.takeIf { (_, overallViolationCount) ->
+        overallViolationCount > 0U
+    }?.let { (formattedViolations, overallViolationCount) ->
+        throw ContractViolationException(
+            overallViolationCount,
+            "The contract input was invalid - $overallViolationCount unique violations were found: $formattedViolations"
+        )
     }
+
+internal class ContractEnforcementContext {
+    /**
+     * A [ContractViolationMap] which can be updated by calls to [requireThat].
+     */
+    private val violations: ContractViolationMap = mutableMapOf()
```

Comment:
I wanted to avoid having to "complicate" things with an internal class, but while developing "glass box" tests for this file, I found the typealias had to be a `MutableMap` for the `checksBody` variant of `validateRequirements` to work, so I just added this wrapping class to ensure safe data manipulation.